DIPs/DIP27.md

```diff
@@ -0,0 +1,172 @@
+# First class functions
+
+| Section        | Value                                  |
+|----------------|----------------------------------------|
+| DIP:           | 27                                     |
+| Author:        | Amaury SÉCHET                          |
+| Status:        |                                        |
+
+## Abstract
+
+This DIP is the first of 3. It intends to define what is a function in D. The DIP propose a radically simple approach by deciding that all function are first class function. The DIP proposes ways to mitigate breakage involved and allow optional parenthesis when no confusion is possible.
+
+## Description
+
+Function are still defined with the same syntax :
+
+```d
+ReturnType functionName(Parameters) {
+  // Function body.
+}
+```
+
+However, this is now strictly equivalent to :
+
+```d
+enum functionName = ReturnType function(Parameters) {
+  // Function body.
+};
+```
+
+The function has a function type and can be used as this anywhere a function is expected :
+
+```d
+static assert(is(typeof(functionName) == ReturnType function(Parameters)); // OK
+
+ReturnType function(Parameters) foo = functionName; // OK
+
+void buzz(ReturnType function(Parameters) qux);
+buzz(foo); // OK
+buzz(bar); // OK
+
+auto a = functionName;
+static assert(is(typeof(a) == ReturnType function(Parameters)); // OK
+
+auto b = functionName();
+static assert(is(typeof(b) == ReturnType); // OK
+```
+
+Note that what is above is simply consequences of the simple rule expressed above.
+
+### Is Expression
+
+The function type specialization used in is expression has to be modified as well. It does match the first class function we defined above, and alias parameters as defined on https://dlang.org/spec/expression.html#IsExpression .
+
+```d
+void foo() {}
+auto bar = foo;
+
+static assert(is(typeof(foo) == function)); // OK
+static assert(is(typeof(bar) == function)); // OK
+static assert(is(void function() == function)); // OK
+```
+
+Matching parameters using an identifier is also affected to work with all first class functions.
+
+```d
+static assert(is(typeof(foo) P == function)); // OK, P is an empty sequence.
+static assert(is(void function(uint) P == function)); // OK, P is an a sequence of one element : uint.
+```
+
+Alterantively, this behavior may be dropped as '''__parameters''' already provides it.
+
+### Optional parentheses
+
+Redundant parenthesis can be a burden for argument-less function class (or single argument function calls used as UFCS). Implicit function call is performed is the following cases :
```

Comment:
Typo: Parentheses

**An** implicit function call is performed **as** in the following cases